base/v0_5_exp/translate.go

```diff
@@ -218,6 +220,155 @@ func translateLink(from Link, options common.TranslateOptions) (to types.Link, t
 	return
 }
 
+func translatePasswdUser(from PasswdUser, options common.TranslateOptions) (to types.PasswdUser, tm translate.TranslationSet, r report.Report) {
+	tr := translate.NewTranslator("yaml", "json", options)
+	tm, r = translate.Prefixed(tr, "gecos", &from.Gecos, &to.Gecos)
+	translate.MergeP(tr, tm, &r, "groups", &from.Groups, &to.Groups)
+	translate.MergeP2(tr, tm, &r, "home_dir", &from.HomeDir, "homeDir", &to.HomeDir)
+	translate.MergeP(tr, tm, &r, "name", &from.Name, &to.Name)
+	translate.MergeP2(tr, tm, &r, "no_create_home", &from.NoCreateHome, "noCreateHome", &to.NoCreateHome)
+	translate.MergeP2(tr, tm, &r, "no_log_init", &from.NoLogInit, "noLogInit", &to.NoLogInit)
+	translate.MergeP2(tr, tm, &r, "no_user_group", &from.NoUserGroup, "noUserGroup", &to.NoUserGroup)
+	translate.MergeP2(tr, tm, &r, "password_hash", &from.PasswordHash, "passwordHash", &to.PasswordHash)
+	translate.MergeP2(tr, tm, &r, "primary_group", &from.PrimaryGroup, "primaryGroup", &to.PrimaryGroup)
+	translate.MergeP2(tr, tm, &r, "should_exist", &from.ShouldExist, "shouldExist", &to.ShouldExist)
+	translate.MergeP(tr, tm, &r, "shell", &from.Shell, &to.Shell)
+	translate.MergeP(tr, tm, &r, "system", &from.System, &to.System)
+	translate.MergeP(tr, tm, &r, "uid", &from.UID, &to.UID)
+
+	var c path.ContextPath
+	if from.SSHAuthorizedKeys != nil && len(from.SSHAuthorizedKeys) > 0 {
+		c = path.New("yaml", "ssh_authorized_keys")
+		tm.AddTranslation(c, path.New("json", "sshAuthorizedKeys"))
+		for i, sshKey := range from.SSHAuthorizedKeys {
+			sshKey = SSHAuthorizedKey(strings.TrimSpace(string(sshKey)))
+			if len(sshKey) == 0 {
+				r.AddOnError(c, common.ErrSSHKeyEmpty)
```

Comment:
Will change it to a warning – I don't think adding an empty SSH key can be intended behaviour; omitting the whole `ssh_authorized_keys` would be the way to go for that.

Reply:
Well, there are many fields that don't make sense empty, but we don't systematically check for them, so warning in this case seems a bit odd.  Is there a particular concern about empty keys?

(Even if we did do the check, it should be during the validation pass, since we don't need any information from the translation pass in order to do it.)

Reply:
That's a fair point and one I actually didn't consider yet. Based on that I am personally still in favour of giving a better user experience by informing (by warning, error, or else) users of (potentially) unintended behaviour, however, I agree that this should go then and be replaced with `MergeP2`.
As a follow up, I would propose a separate (tracking/discussion) issue be opened concerning technical and semantic validation of user inputs which I erroneously tried to achieve here.
My concern for empty SSH keys is two-fold: there already is a way to have an "empty" one, simply omit it in the butane file. A second way to achieve the same result strikes me as a bad design choice where there is no real upside to it I can see. Secondly, putting an empty SSH key strikes me as way for users to forget actually replacing a blank placeholder – particularly a file – with wanted contents. After all the SSH keys are (usually) the only way to connect to a (remote) system easily. It'd be a shame if you just rolled out a system and couldn't connect because the config system didn't point out a potential but obvious error beforehand, no?